text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+
+
+## Risks and Drawbacks
+- FOLIO's multi-versioning scheme involving tenants using different versions of modules will cause functional issues. Multiple versions of a module will be consuming from the same topic and there are no guarantees about which module version will consume a message sourced from an incompatible version.
+- Insights like "how many messages have yet to be processed for a tenant" will be harder to derive.
+- Per tenant topic settings will not possible.
+- [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw) will no longer be possible. This solution is meant to be temporary as its name implies, a stop gap to a permanent solution. Creating topics for each tenant has been implemented and lightly enforced via the [folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper) library. FOLIO is not able to enforce tenant security and isolation within its boundaries: the other two parts are not defined within FOLIO code, it is up to the hosting provider to implement. 
+Kafka Topics per tenant is one part of a three part proposal to imbibe FOLIO tenant isolation in Kafka. More details about this proposal, [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw). Only one of these items was completed.
+
+## Related Concerns
+
+### Finishing Temporary Kafka Security Solution
+Attempting to implement the other two parts of the [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw) would probably force a decision to have dedicated modules instances to retain distinct credentials needed to access Kafka thereby raising FOLIO's cost-to-host. Having modules per tenant is cost prohibitive, causes inefficient use of some resources and overwhelm other resources e.g. increasing database connections for each module instance. At the time of writing, FOLIO's own release infrastructure does not have modules per tenant or Kafka security configurations as described by the Temporary Kafka Security Solution. Another consequence would be the need for a process to create Kafka users when a tenant is installed in okApi. A representation of Kafka ACLs would also need to be created for FOLIO developers to modify definitions for which modules can access any particular topic. These items are not defined by the temporary solution and is left for a hosting provider to figure out.
+
+### Supporting Module Multi-Versioning Via Topic Versioning
+FOLIO's multi-versioning scheme is still a gap not covered by this change. An approach could be to create topics for each module version. This will allow specific messages to be delivered to specific versions without causing a partition count explosion when a new tenant is installed. Topics belonging to older modules versions can be removed safely.
```

Comment:
No, you have interpreted it correctly. The ultimate push to reduce the number of partitions  would be to have only one topic that every FOLIO module will use and then having multiple event types. This would be very undesirable and an over-optimization towards fixing the problem, I do not recommend this. Having a topic per module version will have less partition count than having a topic per tenant, in general(we don’t expect 10 versions running concurrently but we expect 10 tenants). The goal of the RFC is the reduce the partition counts, but it needs to be balanced with other factors as well. Here is the order of partition count

One topic per FOLIO cluster < Topic per application concern < Topic per app concern per module version < Topic per app concern per tenant

This section of the document is meant to demonstrate that multi-versioning has been analyzed but ultimately I am opting not to not address due to the complexities described.

Reply:
> This section of the document is meant to demonstrate that multi-versioning has been analyzed but ultimately I am opting not to not address due to the complexities described.

Ok, I think it needs to be made really explicit that this is not going to be addressed and why that is considered acceptable.

Reply:
We should leave this thread unresolved while other threads on multi-versioning is concluded. I will add our decision to the doc explicitly.

Reply:
multi-versioning included in the open question section of the doc.